File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        def size(root, sizes):
            if root:
                l_size = size(root.left, sizes)
                r_size = size(root.right, sizes)
                cur_size = l_size + r_size + 1
                sizes[root.val] = cur_size
                return cur_size
            else:
                return 0
        sizes = {}
        size(root, sizes)

        def kth(root, k):
            l_size = 0 if root.left is None else sizes[root.left.val]
            if l_size == k - 1:
                return root.val
            elif l_size > k - 1:
                return kth(root.left, k)
            else:
                return kth(root.right, k - l_size - 1)
        return kth(root, k)

File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-0.java
class Solution {
    public int findMin(int[] nums) {
        return recur(nums, 0, nums.length - 1);
    }

    public int recur(int[] nums, int left, int right) {
        int mid = (left + right) / 2;
        if (nums[left] <= nums[mid]) {
            if (nums[mid] <= nums[right]) {
                return nums[left];
            } else {
                return recur(nums, mid + 1, right);
            }
        } else {
            return recur(nums, left, mid);
        }
    }
}
